src/main/scala/viper/gobra/frontend/Desugar.scala

```diff
@@ -1126,6 +1126,174 @@ object Desugar {
 
           case n@PBreak(label) => unit(in.Break(label.map(x => x.name), nm.fetchBreakLabel(n, info))(src))
 
+          case n@PShortForRange(range, shorts, spec, body) =>
+            // is a block as it introduces new variables
+            unit(block(for {
+              exp <- goE(range.exp)
+              (elems, typ) = underlyingType(exp.typ) match {
+                case slice : in.SliceT => (slice.elems, slice)
+                case array : in.ArrayT => (array.elems, array)
+                case _ : in.MapT => violation("Maps are not supported yet in range")
+                case _ : in.StringT => violation("Strings are not supported yet in range")
+                case t => violation(s"Range not applicable to type $t")
+              }
+
+              // index is the place where we store the indices of the range expression
+              indexLeft <- leftOfAssignmentD(shorts(0))(in.IntT(Addressability.exclusiveVariable))
+              indexVar = in.Assignee.Var(indexLeft)
+              indexSrc = meta(shorts(0))
+              rangeSrc = meta(range)
+              indexAss = singleAss(indexVar, in.IntLit(0)(rangeSrc))(indexSrc)
+              incrIndex = singleAss(indexVar, in.Add(indexLeft, in.IntLit(1)(indexSrc))(indexSrc))(indexSrc)
+
+              rangeExpSrc = meta(range.exp)
+
+              // in go the range expression is only computed once before the iteration begins
+              // we do that by storing it in copiedVar
+              copiedVar <- freshDeclaredExclusiveVar(exp.typ, n, info)(rangeExpSrc)
+              copyAss = singleAss(in.Assignee.Var(copiedVar), exp)(rangeExpSrc)
+
+              // get the length of the expression
+              length = in.Length(copiedVar)(rangeSrc)
+              cond = in.LessCmp(indexLeft, length)(rangeSrc)
+
+              // this invariant states that the index is between 0 and the length of the array/slice (both inclusive)
+              indexBoundsInv = in.ExprAssertion(in.And(
+                in.AtMostCmp(in.IntLit(0)(rangeSrc), indexLeft)(rangeSrc),
+                in.AtMostCmp(indexLeft, length)(rangeSrc))(rangeSrc))(rangeSrc)
+
+              // standard for loop desugaring
+              (dInvPre, dInv) <- prelude(sequence(spec.invariants map assertionD(ctx)))
+              (dTerPre, dTer) <- prelude(option(spec.terminationMeasure map terminationMeasureD(ctx)))
+
+              dBody = blockD(ctx)(body)
+
+              continueLabelName = nm.continueLabel(n, info)
+              continueLoopLabelProxy = in.LabelProxy(continueLabelName)(src)
+              continueLoopLabel = in.Label(continueLoopLabelProxy)(src)
+
+              breakLabelName = nm.breakLabel(n, info)
+              breakLoopLabelProxy = in.LabelProxy(breakLabelName)(src)
+              _ <- declare(breakLoopLabelProxy)
+              breakLoopLabel = in.Label(breakLoopLabelProxy)(src)
+
+              wh = if (shorts.length == 2) {
```

Comment:
might not be immediately obvious here why we do this case analysis. I would add a very small skeleton of a loop where this case is supposed to apply (in comments) and I would do the same for the `else` case